soc/arm/atmel_sam0/common/power.c

```diff
@@ -0,0 +1,88 @@
+/*
+ * Copyright (c) 2019 Benjamin Valentin
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+#include <zephyr.h>
+#include <power.h>
+#include <soc.h>
+
+#include <logging/log.h>
+LOG_MODULE_DECLARE(soc, CONFIG_SOC_LOG_LEVEL);
+
+static inline void sam0_idle(uint8_t mode)
+{
+	SCB->SCR &= ~SCB_SCR_SLEEPDEEP_Msk;
+	PM->SLEEP.reg = mode;
+	__DSB();
+	__WFI();
+}
+
+static inline void sam0_deep_sleep(void)
+{
+	SCB->SCR |= SCB_SCR_SLEEPDEEP_Msk;
+	__DSB();
+	__WFI();
+}
+
+/* Invoke Low Power/System Off specific Tasks */
+void sys_set_power_state(enum power_states state)
+{
```

Comment:
Nit: Errata 1.14.2 on SAMD21 (wake failures) silicon revision <= C (DSU->DID.bit.REVISION <= 0x2) should probably be accounted for here (i.e. write `NVMCTRL->CTRLB.bit.SLEEPPRM = NVMCTRL_CTRLB_SLEEPPRM_DISABLED_Val`).  I have no idea about if those are actually in the wild or just on paper in the errata, so I don't know if that's really required.

Reply:
I've added a check for this errata.